### DISCUSSION TITLE
[bug?] elevation from custom geotiff is weirdly cut

### DISCUSSION THREAD

**Discussion by Mis012** [osmandapp/OsmAnd#20543] [General]
### Description

I reprojected a geotiff of the whole Czech Republic so that OsmAnd can use it for elevation with the OpenGL renderer, however instead of the elevation being applied in the whole Czech Republic, tiles within a certain distance from the country border (equivalent to the edge of the elevation data in this case) don't have elevation applied to them. It's perhaps easier to see this when the geotiff is displayed as a map layer instead of being used for elevation (numpad 9 in `bird`, undocumented for some reason): 
![the issue at hand visible in bird, with the country border highlighted](https://github.com/user-attachments/assets/0116febf-4573-4edb-b231-c87d47b63f70)
(I underlayed a classic map (numpad 1) to highlight where the country border is)

this is what the geotiff looks like in QGIS, seems completely fine.
![the geotiff in QGIS](https://github.com/user-attachments/assets/8f6306ff-a28a-4dd4-b6bd-e5bbe8adbf73)
### 

### Steps to reproduce

get the file I was testing with (I made it myself by reprojecting the original, since OsmAnd is picky about the projection used; I wanted to share it via gitlab since the original is CC-BY-SA 4.0 and other people might find this useful, but it seems it's too large for that, so please download it in the next 30 days; original is at https://atom.cuzk.cz/)
send.cyberjake.xyz/download/db80a926d7a6c629/#mLjPpYXoHZD-mC5gMbSsTw

### Actual result

tiles within a certain distance from the country border (equivalent to the edge of the elevation data in this case) don't have elevation applied to them.

### Expected result

The elevation data is applied everywhere where the geotiff has a value other than NODATA

### Your Environment (required)

```
OsmAnd Version: latest master (`repo info` below) (screenshots are from older version but I confirmed the issue is still present)
Android/iOS version: issue can be reproduced in `bird`
Device model: x86 laptop running Linux
Crash-Logs: no crash
```

`repo info`:
```
Manifest branch: master
Manifest merge branch: refs/heads/master
Manifest groups: default,platform-linux
----------------------------
Project: osmandapp/Osmand.git
Current revision: b9e0148238a2585f3ce4b4558f2b66f6cb416735
Manifest revision: master
Local Branches: 0
----------------------------
Project: osmandapp/OsmAnd-build.git
Current revision: 2108cd6a59496ca78f59feb4d6378c988a7a8da6
Manifest revision: master
Local Branches: 0
----------------------------
Project: osmandapp/OsmAnd-core.git
Current revision: cedbd8b6c85956f36f470e4e213ab0190777e4cb
Current branch: master
Manifest revision: master
Local Branches: 2 [r4.6, master]
----------------------------
Project: osmandapp/OsmAnd-rendering-tests.git
Current revision: 7d5913b5717810c3f251b9ac0ace38d5dfa3fb21
Manifest revision: master
Local Branches: 0
----------------------------
Project: osmandapp/OsmAnd-resources.git
Current revision: 988d4af49900ed83f689c00aa6eafa6eea9574f4
Manifest revision: master
Local Branches: 0
----------------------------
Project: osmandapp/OsmAnd-tools.git
Current revision: 22fe7eb9ad06521eb1793b1988514d51999153a0
Manifest revision: master
Local Branches: 0
----------------------------
```

**Comment by mariush444** [comment]
Off topic a little
@Mis012 
Can you please share info step by step in detailes how you create geotiff for osmand? I got lost in "fighting" with it. 
Additionally  I interested in "numpad 9 in bird, undocumented for some reason". Thank you in advance.

**Comment by Mis012** [comment]
I just took the existing geotiff which was using a different projection and ran this: `gdalwarp -t_srs 'WGS84' -wo OPTIMIZE_SIZE=YES -wo GDAL_CACHEMAX=500 -wo NUM_THREADS=6 -wm 500 -co COMPRESS=LZW -multi -co BIGTIFF=YES EL-GRID.tif EL-GRID_WGS84.tif`

the `bird` tool has these options for tile sources, selected using the numpad:
![image](https://github.com/user-attachments/assets/75b13390-65d6-448f-901c-6b4b48bcf91c)

in the code, there's additionally a case for `9`:
```C++
    else if (idx == 9)
    {
        if (geotiffDirSpecified)
        {
            if (!hillshade)
            {
                if (!geotiffCollection)
                {
                    const auto manualGeoTiffCollection = new OsmAnd::GeoTiffCollection();
                    manualGeoTiffCollection->addDirectory(geotiffDir);
                    manualGeoTiffCollection->setLocalCache(cacheDir);
                    geotiffCollection.reset(manualGeoTiffCollection);
                }
                auto reliefRasterMapLayerProvider = new OsmAnd::HillshadeRasterMapLayerProvider(geotiffCollection,
                    QString("../../../tools/obf-generation/heightmap/color/hillshade_main.txt"),
                    QString("../../../tools/obf-generation/heightmap/color/color_slope.txt"));
                //auto reliefRasterMapLayerProvider = new OsmAnd::SlopeRasterMapLayerProvider(geotiffCollection,
                //    QString("../../../../tools/obf-generation/heightmap/color/slopes_main.txt"));

                renderer->setMapLayerProvider(layerIdx,
                    std::shared_ptr<OsmAnd::IMapLayerProvider>(reliefRasterMapLayerProvider));
                hillshade = true;
            }
            else
            {
                if (!heightmap)
                    geotiffCollection.reset();
                renderer->resetMapLayerProvider(layerIdx);
                hillshade = false;
            }
        }
    }
```

**Reply by Mis012** [reply]
actually I think `TILED=YES` is missing here, that was also important

**Comment by vshcherb** [comment]
You need to convert into Mercator projection 14th zoom if I remember correctly. https://github.com/osmandapp/OsmAnd-tools/blob/master/obf-generation/heightmap/generate_tile.sh#L176

**Reply by Mis012** [reply]
well, it looks completely fine except for where it's close to the edge...
also, I obviously don't want to change the resolution, since that would kill the point of using a higher resolution source (I'm quite fine with it taking up 7-8GiB for the improvement in resolution)

**Reply by Mis012** [reply]
at least it seems to me that all setting `-tr` according to that formula would do is reduce the resolution, and it seems that OsmAnd mostly handles the higher resolution fine (though `bird` seems to have some LOD artifacts, thankfully everything works fine in the android app)

everything looks fine...
![image](https://github.com/user-attachments/assets/c793f755-8ad2-472d-9e50-679341d4cd74)

...except near the edges
![image](https://github.com/user-attachments/assets/e7290a7b-644d-447a-895e-be0e76159305)


btw, my geotiff is not cut into tiles, but the geotiffs downloaded by OsmAnd are not cut into tiles either, which seems to be a non-negotiable part of that script...?

**Reply by vshcherb** [reply]
Cutting to tiles is not needed, near edges don't work that's correct. We left that artifact so you could see the boundaries of 3D map

**Reply by Mis012** [reply]
but the boundary is supposed to be at the country border (as seen on the QGIS screenshot), not kilometers away from it...

**Reply by vshcherb** [reply]
Internally everything is retrieved as tiles (14th zoom I guess), if tile is not complete or couldn't be parsed from Geotiff it's not displayed

**Reply by Mis012** [reply]
a geotiff always has to be rectangular obviously, the pixels where no data is available are set as NODATA, which I assume is what the "official" geotiffs do as well? Maybe the NODATA value is hardcoded and it's different between the two?

**Reply by Mis012** [reply]
hm, I see... it doesn't
![image](https://github.com/user-attachments/assets/21e56e55-3b6f-40ba-8aa8-ab44efcfd524)

isn't that wasteful? it means you will have stuff duplicated when you download geotiffs for two neighboring countries

**Comment by Mis012** [comment]
So the issue is that any tile with even a single NODATA pixel gets rejected? Can that be fixed? I guess I could work around this by using `gdal_merge` to overlay my data over the data from the stock geotiff, even though that seems inelegant...

**Reply by Andrewschenko** [reply]
For good performance, we need to easily know if a particular tiff has complete data for a particular tile. That's why we provide heightmap tiffs that contain only full tiles of ninth zoom level.

**Reply by Mis012** [reply]
hm, I see... maybe there should be an error message if there is a tile which is considered invalid, though if there was some documentation regarding making custom geotiffs and that documentation was easy to find then could maybe just put this information in there? Actually, a text file that gets deposited in the geotiff directory would be pretty hard to miss, and that could contain a link to the documentation

**Reply by Andrewschenko** [reply]
What do you mean by "considered invalid"?

**Reply by Mis012** [reply]
well, the tiles which had parts of them as NODATA didn't get displayed at all

**Reply by Andrewschenko** [reply]
If any pixel in the corners or center of a tile has no data, then we do not display that tile elevated. That's why we need only either completely "full" tiles or completely "empty" ones in the tiff file.

**Reply by Mis012** [reply]
yeah, this would be nice to document, as part of the currently non-existent documentation on how to make a compatible geotiff

**Comment by Mis012** [comment]
so the following worked, although the interpolation for the parts that had to be scaled is sad:
`gdalbuildvrt -resolution highest EL-GRID_WGS84_fulltiles.vrt ~/Downloads/Heightmap\ Czech-republic\ europe.tif EL-GRID_WGS84.tif`
`gdal_translate -co TILED=YES -co NUM_THREADS=8 -co COMPRESS=LZW -co BIGTIFF=YES -co SPARSE_OK=TRUE EL-GRID_WGS84_fulltiles.vrt EL-GRID_WGS84_fulltiles.tif`

also, at least with `bird` it seems that ZSTD works nicely even at default compression level (6GiB vs 8.6GiB with LZW) but the decompression feels noticeably slower :/   
Although that doesn't make any sense, because ZSTD decompression should surely be faster than LZW...
